hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCancelAirdropHandler.java

```diff
@@ -126,15 +134,127 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
+    }
+
+    private static void configValidation(Configuration configuration, TokenCancelAirdropTransactionBody op) {
+        var tokensConfig = configuration.getConfigData(TokensConfig.class);
         validateTrue(tokensConfig.cancelTokenAirdropEnabled(), NOT_SUPPORTED);
+        validateFalse(
+                op.pendingAirdrops().size() > tokensConfig.maxAllowedPendingAirdropsToCancel(),
+                MAX_PENDING_AIRDROP_ID_EXCEEDED);
+    }
+
+    private static void validatePendingAirdropIds(
+            HandleContext context,
+            List<PendingAirdropId> pendingAirdropIds,
+            AccountID payer,
+            WritableAccountStore accountStore,
+            WritableAirdropStore airdropStore) {
+        final var tokenStore = context.storeFactory().readableStore(ReadableTokenStore.class);
+        final var nftStore = context.storeFactory().readableStore(ReadableNftStore.class);
+
+        for (var pendingAirdropId : pendingAirdropIds) {
+            validateTrue(payer.equals(pendingAirdropId.senderIdOrThrow()), INVALID_ACCOUNT_ID);
+            if (pendingAirdropId.hasFungibleTokenType()) {
+                getIfUsable(pendingAirdropId.fungibleTokenTypeOrThrow(), tokenStore);
+            } else {
+                final var nft = pendingAirdropId.nonFungibleTokenOrThrow();
+                validateTrue(nftStore.get(nft) != null, INVALID_NFT_ID);
+            }
+            getIfUsable(
+                    pendingAirdropId.senderIdOrThrow(), accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+            getIfUsable(
+                    pendingAirdropId.receiverIdOrThrow(), accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+            validateTrue(airdropStore.exists(pendingAirdropId), INVALID_TRANSACTION_BODY);
+        }
+    }
+
+    private static void deleteAccountAirdropsFromAccount(
+            List<PendingAirdropId> pendingAirdropIds,
+            WritableAirdropStore airdropStore,
+            Account payerAccount,
+            WritableAccountStore accountStore) {
+        for (var pendingAirdropsToCancel : pendingAirdropIds) {
+            final var accountAirdropToCancel = airdropStore.getForModify(pendingAirdropsToCancel);
+            validateTrue(accountAirdropToCancel != null, INVALID_TRANSACTION_BODY);
+
+            // Update the account to point to the new head if the current one is being cancelled
+            if (pendingAirdropsToCancel.equals(payerAccount.headPendingAirdropId())) {
+                final var updatedPayer = payerAccount
+                        .copyBuilder()
+                        .headPendingAirdropId(accountAirdropToCancel.nextAirdrop())
+                        .build();
+                accountStore.put(updatedPayer);
+            }
+
+            final var prevAirdropId = accountAirdropToCancel.previousAirdrop();
+            final var nextAirdropId = accountAirdropToCancel.nextAirdrop();
```

Comment:
We should probably re-use this logic between claim and cancel. Will be nice to have all pointers change logic in one place and re-used

Reply:
Yeah, we discussed that with Jivko. He extended this logic into a new class that I will copy-paste in this PR.

Reply:
I added a `PendingAirdropUpdater` class 👍